File: lib/graph/bipariate_match.cpp
struct bipariate_match {
  int n,m;
  vector<int> mt;
  vector<vector<int>> adj;
  bipariate_match(int n, int m) : n(n), m(m), mt(m, -1), adj(n, vector<int>()) {}
  void add(int u, int v) {
    adj[u].push_back(v);
  }
  void run() {
    vector<bool> vis;
    auto dfs = [&](auto self, int u) -> bool {
      if (vis[u]) return false;
      vis[u] = true;
      for (int v : adj[u]) {
        if (mt[v] == -1 || self(self, mt[v])) {
          mt[v] = u;
          return true;
        }
      }
      return false;
    };

    // arbitrary matching heuristics
    vector<bool> used(n);
    for (int i=0;i<n;i++){
      for (int v : adj[i]) {
        if (mt[v] == -1) {
          mt[v] = i;
          used[i] = true;
          break;
        }
      }
    }

    for (int i=0;i<n;i++) {
      if (used[i]) continue; // add mt[i] != -1 check when abitrary
      vis.assign(n, false);
      dfs(dfs, i);
    }
  }
};

File: lib/graph/bipariate_match.py
class bipariate_match:
  def __init__(self, n, m):
    self.n = n
    self.m = m
    self.mt = [-1] * m
    self.adj = [[] for i in range(n)]
  def add(self, u, v):
    self.adj[u].append(v)
  def run(self):
    used = [False] * self.n
    for i in range(self.n):
      for v in self.adj[i]:
        if self.mt[v] == -1:
          self.mt[v] = i
          used[i] = True
          break
    def dfs(vis, u):
      if vis[u]:
        return False
      vis[u] = True
      for v in self.adj[u]:
        if self.mt[v] == -1 or dfs(vis, self.mt[v]):
          self.mt[v] = u
          return True
      return False
    for i in range(self.n):
      # add mt[i] == -1 check if arbitrary
      if used[i]:
        continue
      vis = [False] * self.n
      dfs(vis, i)